port/linux/ipadapter.c

```diff
@@ -933,23 +1017,31 @@ to_timeval(oc_clock_time_t ticks)
 }
 #endif /* OC_HAS_FEATURE_TCP_ASYNC_CONNECT */
 
-static void *
-network_event_thread(void *data)
+static void
+add_control_flow_rfds(fd_set *output_set, const ip_context_t *dev)
 {
-  ip_context_t *dev = (ip_context_t *)data;
-  FD_ZERO(&dev->rfds);
   /* Monitor network interface changes on the platform from only the 0th
    * logical device
    */
   if (dev->device == 0) {
-    FD_SET(g_ifchange_sock, &dev->rfds);
+    FD_SET(g_ifchange_sock, output_set);
   }
-  FD_SET(dev->shutdown_pipe[0], &dev->rfds);
+  FD_SET(dev->wakeup_pipe[0], output_set);
+}
+
+static void *
+network_event_thread(void *data)
+{
+  ip_context_t *dev = (ip_context_t *)data;
+  FD_ZERO(&dev->rfds);
 
   udp_add_socks_to_rfd_set(dev);
+  add_control_flow_rfds(&dev->rfds, dev);
 #ifdef OC_TCP
   tcp_add_socks_to_rfd_set(dev);
+  tcp_add_controlflow_socks_to_rfd_set(&dev->rfds, dev);
 #endif /* OC_TCP */
+  int max_read_fd = fds_max(&dev->rfds);
```

Comment:
It needs to be calculated under while cycle `while (OC_ATOMIC_LOAD8(dev->terminate) != 1)` because new TCP connections can increase value.

Also calculate  it from fds in dev instance - you don't need to iterate up to FD_SETSIZE

Reply:
Done. The max value of file descriptor is calculated from the recent device descriptor set. It only recalculates when the set of descriptors changes, so iterating up to the value of FD_SETSIZE should not be a performance issue given that sockets do not change frequently.